### PR TITLE
Refactoring cosmos client, options  and adding user

### PR DESCRIPTION
**Pull request by mbhaskar**
Refactored CosmosClient to use builder pattern and removed CosmosConfiguration
CosmosDatabase, Container and Item Requestoptions use composition now
Adding CosmosUser API to database

---
This change is [Reviewable](https://reviewable.io/reviews/azure/azure-cosmosdb-java/139)

### CODE REVIEW

File: sdk/src/main/java/com/microsoft/azure/cosmos/CosmosClient.java [hunk 1]
```diff
@@ -35,49 +38,170 @@
...
+
+    }
+
+    /**
+     * Get the service endpoint
+     * @return the service endpoint
+     */
+    public String getServiceEndpoint() {
+        return serviceEndpoint;
+    }
```
Comment (JonathanGiles) [inline sdk/src/main/java/com/microsoft/azure/cosmos/CosmosClient.java]: Do we need public getter API in the builder? I would rather it wouldn't be, if possible.

File: sdk/src/main/java/com/microsoft/azure/cosmos/CosmosClient.java [hunk 2]
```diff
@@ -35,49 +38,170 @@
...
+     *         .key(key)
+     *         .connectionPolicy(connectionPolicy)
+     *         .consistencyLevel(ConsistencyLevel.Session)
+     *         .build();
+     * }
+     * </pre>
      */
-    public static CosmosClient create(CosmosConfiguration cosmosConfiguration) {
-        return new CosmosClient(cosmosConfiguration);
+    public static class Builder {
```
[the remark this answers is not in the document]
Comment (mbhaskar) [reply sdk/src/main/java/com/microsoft/azure/cosmos/CosmosClient.java]: Done

File: sdk/src/main/java/com/microsoft/azure/cosmos/CosmosClient.java (hunk 1 above)
Comment (mbhaskar) [reply sdk/src/main/java/com/microsoft/azure/cosmos/CosmosClient.java]: These are part of CosmosClient

**Comment by mbhaskar** [comment]
> > Reviewed 19 of 19 files at r1.
> > _[Reviewable](https://reviewable.io/reviews/azure/azure-cosmosdb-java/139)_ status: all files reviewed, 5 unresolved discussions (waiting on @christopheranderson, @kushagraThapar, and @mbhaskar)
> 
> _[sdk/src/main/java/com/microsoft/azure/cosmos/CosmosClient.java, line 71 at r1](https://reviewable.io/reviews/azure/azure-cosmosdb-java/139#-Lfz9B3LCPe19HWvGVxc:-Lfz9B3LCPe19HWvGVxd:be4afbd) ([raw file](https://github.com/azure/azure-cosmosdb-java/blob/ee0fa0909f4ca4068b9176484845c93806dbfaad/sdk/src/main/java/com/microsoft/azure/cosmos/CosmosClient.java#L71)):_
> 
> > ```
> > ········
> > ```
> 
> nit: spaces
> 
> _[sdk/src/main/java/com/microsoft/azure/cosmos/CosmosClient.java, line 194 at r1](https://reviewable.io/reviews/azure/azure-cosmosdb-java/139#-Lfz9ipa58SaUchDD9xg:-Lfz9ipa58SaUchDD9xh:biowumh) ([raw file](https://github.com/azure/azure-cosmosdb-java/blob/ee0fa0909f4ca4068b9176484845c93806dbfaad/sdk/src/main/java/com/microsoft/azure/cosmos/CosmosClient.java#L194)):_
> 
> Quoted 11 lines of code…
> why some of the getters here are pubic and some other getters are not public?
> 
> _[sdk/src/main/java/com/microsoft/azure/cosmos/CosmosContainer.java, line 538 at r1](https://reviewable.io/reviews/azure/azure-cosmosdb-java/139#-LfzA1_L9mAKnap90yCG:-LfzA1_L9mAKnap90yCH:b-9ye2ef) ([raw file](https://github.com/azure/azure-cosmosdb-java/blob/ee0fa0909f4ca4068b9176484845c93806dbfaad/sdk/src/main/java/com/microsoft/azure/cosmos/CosmosContainer.java#L538)):_
> 
> > ```
> > public CosmosTrigger getTrigger(String id){
> > ```
> 
> I thought we will be exposing stored proc only as public API in v3 and not triggers. no?
> 
> _[sdk/src/test/java/com/microsoft/azure/cosmos/CosmosTestSuiteBase.java, line 108 at r1](https://reviewable.io/reviews/azure/azure-cosmosdb-java/139#-LfzBGFE8D-DBFXGR8qL:-LfzBGFE8D-DBFXGR8qM:bimtg70) ([raw file](https://github.com/azure/azure-cosmosdb-java/blob/ee0fa0909f4ca4068b9176484845c93806dbfaad/sdk/src/test/java/com/microsoft/azure/cosmos/CosmosTestSuiteBase.java#L108)):_
> 
> > ```
> > ········
> > ```
> 
> nit: empty spaces
> 
> _[sdk/src/test/java/com/microsoft/azure/cosmosdb/rx/internal/TestSuiteBase.java, line 241 at r1](https://reviewable.io/reviews/azure/azure-cosmosdb-java/139#-LfzBXNk5eRmVYFFlo0o:-LfzBXNk5eRmVYFFlo0p:b-y4792y) ([raw file](https://github.com/azure/azure-cosmosdb-java/blob/ee0fa0909f4ca4068b9176484845c93806dbfaad/sdk/src/test/java/com/microsoft/azure/cosmosdb/rx/internal/TestSuiteBase.java#L241)):_
> 
> > ```
> > requestOptions.partitionKey(new PartitionKey(propertyValue));
> > ```
> 
> In this PR you changed APIs of v3 sdk, v2 API remained unchanged. no need to change anything in v2 files.
> undo changes in this file.

Done

**Comment by moderakh** [comment]
There are javadoc warnings:

```
[WARNING] Javadoc Warnings
[WARNING] D:\a\1\s\sdk\src\main\java\com\microsoft\azure\cosmos\CosmosClient.java:281: warning: no description for @return
[WARNING] * @return
[WARNING] ^
[WARNING] D:\a\1\s\sdk\src\main\java\com\microsoft\azure\cosmos\CosmosClientBuilder.java:96: warning: no description for @return
[WARNING] * @return
[WARNING] ^
[WARNING] D:\a\1\s\sdk\src\main\java\com\microsoft\azure\cosmos\CosmosClientBuilder.java:105: warning: no description for @param
[WARNING] * @param connectionPolicy
[WARNING] ^
[WARNING] D:\a\1\s\sdk\src\main\java\com\microsoft\azure\cosmos\CosmosClientBuilder.java:106: warning: no description for @return
[WARNING] * @return
[WARNING] ^
[WARNING] D:\a\1\s\sdk\src\main\java\com\microsoft\azure\cosmos\CosmosContainer.java:71: warning: no @param for options
[WARNING] public Mono<CosmosContainerResponse> read(CosmosContainerRequestOptions options) {
[WARNING] ^
[WARNING] D:\a\1\s\sdk\src\main\java\com\microsoft\azure\cosmos\CosmosContainer.java:299: warning: no @param for partitionKey
[WARNING] public CosmosItem getItem(String id, Object partitionKey){
[WARNING] ^
[WARNING] D:\a\1\s\sdk\src\main\java\com\microsoft\azure\cosmos\CosmosContainer.java:500: warning: no @param for settings
[WARNING] public Mono<CosmosTriggerResponse> createTrigger(CosmosTriggerSettings settings,
[WARNING] ^
[WARNING] D:\a\1\s\sdk\src\main\java\com\microsoft\azure\cosmos\CosmosContainerRequestOptions.java:51: warning: no @return
[WARNING] public CosmosContainerRequestOptions offerThroughput(Integer offerThroughput) {
[WARNING] ^
[WARNING] D:\a\1\s\sdk\src\main\java\com\microsoft\azure\cosmos\CosmosContainerRequestOptions.java:74: warning: no @return
[WARNING] public CosmosContainerRequestOptions populateQuotaInfo(boolean populateQuotaInfo) {
[WARNING] ^
[WARNING] D:\a\1\s\sdk\src\main\java\com\microsoft\azure\cosmos\CosmosContainerRequestOptions.java:93: warning: no @return
[WARNING] public CosmosContainerRequestOptions consistencyLevel(ConsistencyLevel consistencyLevel) {
[WARNING] ^
[WARNING] D:\a\1\s\sdk\src\main\java\com\microsoft\azure\cosmos\CosmosContainerRequestOptions.java:112: warning: no @return
[WARNING] public CosmosContainerRequestOptions sessionToken(String sessionToken) {
[WARNING] ^
[WARNING] D:\a\1\s\sdk\src\main\java\com\microsoft\azure\cosmos\CosmosRequestOptions.java:49: warning: no @return
[WARNING] public CosmosRequestOptions accessCondition(AccessCondition accessCondition) {
[WARNING] ^
[WARNING] D:\a\1\s\sdk\src\main\java\com\microsoft\azure\cosmos\CosmosContainerSettings.java:107: warning: no description for @param
[WARNING] * @param indexingPolicy
[WARNING] ^
[WARNING] D:\a\1\s\sdk\src\main\java\com\microsoft\azure\cosmos\CosmosDatabase.java:103: warning: no @param for options
[WARNING] public Mono<CosmosDatabaseResponse> delete(CosmosRequestOptions options) {
[WARNING] ^
[WARNING] D:\a\1\s\sdk\src\main\java\com\microsoft\azure\cosmos\CosmosDatabase.java:175: warning: no @param for containerSettings
[WARNING] public Mono<CosmosContainerResponse> createContainerIfNotExists(CosmosContainerSettings containerSettings) {
[WARNING] ^
[WARNING] D:\a\1\s\sdk\src\main\java\com\microsoft\azure\cosmos\CosmosDatabaseRequestOptions.java:47: warning: no @return
[WARNING] public CosmosDatabaseRequestOptions offerThroughput(Integer offerThroughput) {
[WARNING] ^
[WARNING] D:\a\1\s\sdk\src\main\java\com\microsoft\azure\cosmos\CosmosDatabaseResponse.java:43: warning: no description for @return
[WARNING] * @return
[WARNING] ^
[WARNING] D:\a\1\s\sdk\src\main\java\com\microsoft\azure\cosmos\CosmosItemRequestOptions.java:77: warning: no @return
[WARNING] public CosmosItemRequestOptions consistencyLevel(ConsistencyLevel consistencyLevel) {
[WARNING] ^
[WARNING] D:\a\1\s\sdk\src\main\java\com\microsoft\azure\cosmos\CosmosItemRequestOptions.java:96: warning: no @return
[WARNING] public CosmosItemRequestOptions indexingDirective(IndexingDirective indexingDirective) {
[WARNING] ^
[WARNING] D:\a\1\s\sdk\src\main\java\com\microsoft\azure\cosmos\CosmosItemRequestOptions.java:115: warning: no @return
[WARNING] public CosmosItemRequestOptions preTriggerInclude(List<String> preTriggerInclude) {
[WARNING] ^
[WARNING] D:\a\1\s\sdk\src\main\java\com\microsoft\azure\cosmos\CosmosItemRequestOptions.java:134: warning: no @return
[WARNING] public CosmosItemRequestOptions postTriggerInclude(List<String> postTriggerInclude) {
[WARNING] ^
[WARNING] D:\a\1\s\sdk\src\main\java\com\microsoft\azure\cosmos\CosmosItemRequestOptions.java:153: warning: no @return
[WARNING] public CosmosItemRequestOptions sessionToken(String sessionToken) {
[WARNING] ^
[WARNING] D:\a\1\s\sdk\src\main\java\com\microsoft\azure\cosmos\CosmosItemRequestOptions.java:162: warning: no @return
[WARNING] public CosmosItemRequestOptions partitionKey(PartitionKey partitionKey) {
[WARNING] ^
[WARNING] D:\a\1\s\sdk\src\main\java\com\microsoft\azure\cosmos\CosmosTriggerResponse.java:43: warning: no description for @return
[WARNING] * @return
[WARNING] ^
[WARNING] D:\a\1\s\sdk\src\main\java\com\microsoft\azure\cosmos\CosmosTriggerResponse.java:51: warning: no description for @return
[WARNING] * @return
[WARNING] ^
[WARNING] D:\a\1\s\sdk\src\main\java\com\microsoft\azure\cosmos\CosmosUser.java:18: warning: no description for @param
[WARNING] * @param options
[WARNING] ^
[WARNING] D:\a\1\s\sdk\src\main\java\com\microsoft\azure\cosmos\CosmosUser.java:19: warning: no description for @return
[WARNING] * @return
[WARNING] ^
[WARNING] D:\a\1\s\sdk\src\main\java\com\microsoft\azure\cosmos\CosmosUser.java:29: warning: no description for @param
[WARNING] * @param userSettings
[WARNING] ^
[WARNING] D:\a\1\s\sdk\src\main\java\com\microsoft\azure\cosmos\CosmosUser.java:30: warning: no description for @param
[WARNING] * @param options
[WARNING] ^
[WARNING] D:\a\1\s\sdk\src\main\java\com\microsoft\azure\cosmos\CosmosUser.java:31: warning: no description for @return
[WARNING] * @return
[WARNING] ^
[WARNING] D:\a\1\s\sdk\src\main\java\com\microsoft\azure\cosmos\CosmosUser.java:41: warning: no description for @param
[WARNING] * @param options
[WARNING] ^
[WARNING] D:\a\1\s\sdk\src\main\java\com\microsoft\azure\cosmos\CosmosUser.java:42: warning: no description for @return
[WARNING] * @return
[WARNING] ^
[WARNING] D:\a\1\s\sdk\src\main\java\com\microsoft\azure\cosmos\CosmosUserResponse.java:21: warning: no description for @return
[WARNING] * @return
[WARNING] ^
[WARNING] D:\a\1\s\sdk\src\main\java\com\microsoft\azure\cosmos\CosmosUserResponse.java:29: warning: no description for @return
[WARNING] * @return
[WARNING] ^
```

File: sdk/src/main/java/com/microsoft/azure/cosmos/CosmosClient.java [hunk 3]
```diff
@@ -23,10 +23,13 @@
 package com.microsoft.azure.cosmos;
```
Comment (JonathanGiles) [inline sdk/src/main/java/com/microsoft/azure/cosmos/CosmosClient.java]: Just a general observation that these copyright headers aren't what we typically provide in Java. I can suggest the actual text if you are interested in standardising.

File: sdk/src/main/java/com/microsoft/azure/cosmos/CosmosClientBuilder.java [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+    private String keyOrResourceToken;
+    private ConnectionPolicy connectionPolicy;
+    private ConsistencyLevel desiredConsistencyLevel;
+    private List<Permission> permissions;
+
+    CosmosClientBuilder() {
+    }
+
+    /**
+     * The service ednpoint url
```
Comment (JonathanGiles) [inline sdk/src/main/java/com/microsoft/azure/cosmos/CosmosClientBuilder.java]: Typo on 'ednpoint'

File: sdk/src/main/java/com/microsoft/azure/cosmos/CosmosClient.java (hunk 3 above)
Comment (mbhaskar) [reply sdk/src/main/java/com/microsoft/azure/cosmos/CosmosClient.java]: Sure. Created an issue for it #164

**Pull request merged**